From b8dfba3abe762785ebb1c558c927f052cc59df5d Mon Sep 17 00:00:00 2001
From: Haibo Chen <haibo.chen@nxp.com>
Date: Mon, 10 Apr 2017 16:16:20 +0800
Subject: [PATCH 1640/5242] MLK-14539 mmc: sdhci: make no-1-8-v also work for
 DDR52 mode

commit  322b949d4083e800554eac3a4eb4aa0b0d242ac5 from
https://source.codeaurora.org/external/imx/linux-imx.git

MMC SDHCI maintainer Adrian Hunter Introduce SDHCI flags for signal
voltage support and set them based on the supported transfer modes,
except in the case where 3V DDR52 is supported but 1.8V is not.

This patch add the support to make eMMC DDR52 only work at 3.3v when
property 'no-1-8-v' defined.

Signed-off-by: Haibo Chen <haibo.chen@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mmc/host/sdhci.c |    6 ++++--
 include/linux/mmc/host.h |    1 +
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/mmc/host/sdhci.c b/drivers/mmc/host/sdhci.c
index adcaff7..68d3953 100644
--- a/drivers/mmc/host/sdhci.c
+++ b/drivers/mmc/host/sdhci.c
@@ -3769,6 +3769,7 @@ int sdhci_setup_host(struct sdhci_host *host)
 	if (host->quirks2 & SDHCI_QUIRK2_NO_1_8_V) {
 		host->caps1 &= ~(SDHCI_SUPPORT_SDR104 | SDHCI_SUPPORT_SDR50 |
 				 SDHCI_SUPPORT_DDR50);
+		mmc->caps2 |= MMC_CAP2_DDR52_3_3V;
 		/*
 		 * The SDHCI controller in a SoC might support HS200/HS400
 		 * (indicated using mmc-hs200-1_8v/mmc-hs400-1_8v dt property),
@@ -3918,10 +3919,11 @@ int sdhci_setup_host(struct sdhci_host *host)
 		goto unreg;
 	}
 
-	if ((mmc->caps & (MMC_CAP_UHS_SDR12 | MMC_CAP_UHS_SDR25 |
+	if (((mmc->caps & (MMC_CAP_UHS_SDR12 | MMC_CAP_UHS_SDR25 |
 			  MMC_CAP_UHS_SDR50 | MMC_CAP_UHS_SDR104 |
 			  MMC_CAP_UHS_DDR50 | MMC_CAP_1_8V_DDR)) ||
-	    (mmc->caps2 & (MMC_CAP2_HS200_1_8V_SDR | MMC_CAP2_HS400_1_8V)))
+	    (mmc->caps2 & (MMC_CAP2_HS200_1_8V_SDR | MMC_CAP2_HS400_1_8V))) &&
+	    !(mmc->caps2 & MMC_CAP2_DDR52_3_3V))
 		host->flags |= SDHCI_SIGNALING_180;
 
 	if (mmc->caps2 & MMC_CAP2_HSX00_1_2V)
diff --git a/include/linux/mmc/host.h b/include/linux/mmc/host.h
index cca2863..e3839cd 100644
--- a/include/linux/mmc/host.h
+++ b/include/linux/mmc/host.h
@@ -361,6 +361,7 @@ struct mmc_host {
 #define MMC_CAP2_CQE_DCMD	(1 << 24)	/* CQE can issue a direct command */
 #define MMC_CAP2_AVOID_3_3V	(1 << 25)	/* Host must negotiate down from 3.3V */
 #define MMC_CAP2_CD_POST	(1 << 26)	/* post card rescan, let client driver to start */
+#define MMC_CAP2_DDR52_3_3V	(1 << 27)	/* Only supprot eMMC DDR52 at 3.3v */
 
 	int			fixed_drv_type;	/* fixed driver type for non-removable media */
 
-- 
1.7.9.5

